internal/connector/connector_test.go

```diff
@@ -0,0 +1,341 @@
+					"rate_iops":      50,
+					"rate_process":   "poisson",
+					"buffered":       0,
+				},
+			},
+		},
+		"template": {
+			stepID:   "hello-world",
+			location: "arcaflow-plugin-template-python@git+https://github.com/arcalot/arcaflow-plugin-template-python.git@52d1a9559c60a615dbd97114572f16d70fa30b1b",
+			input: map[string]any{
+				"name": "arca lot",
+			},
+		},
+		// no git commit hash is used here, so that the code executes
+		// the 'latest' branch of cliwrapper.GetModulePath()
+		"wait": {
+			stepID:   "wait",
+			location: "arcaflow-plugin-wait@git+https://github.com/arcalot/arcaflow-plugin-wait.git",
+			input: map[string]any{
+				"seconds": "0.5",
+			},
+		},
+	}
+
+	rootDir := "/tmp/multi-module"
+	serializedConfig := map[string]any{
+		"workdir":          rootDir,
+		"modulePullPolicy": "IfNotPresent",
+		"pythonSemver":     "3.0.0",
+	}
+
+	// idempotent test directory creation
+	_ = os.RemoveAll(rootDir)
+	assert.NoError(t, os.MkdirAll(rootDir, os.ModePerm))
+
+	factory := pythondeployer.NewFactory()
+	deployerSchema := factory.ConfigurationSchema()
+	unserializedConfig, err := deployerSchema.UnserializeType(serializedConfig)
+	assert.NoError(t, err)
+
+	pythonPath, err := GetPythonPath()
+	assert.NoError(t, err)
+	unserializedConfig.PythonPath = pythonPath
+
+	// Choose how many connectors and plugins to make
+	const n_connectors = 4
+	const n_plugin_copies = 10
+	wg := sync.WaitGroup{}
+	wg.Add(n_connectors * len(testModules) * n_plugin_copies)
+
+	logger := log.NewTestLogger(t)
+
+	// Test for issues that might occur during concurrent creation of connectors
+	// and deployment of plugins
+	// Make a goroutine for each connector
+	for j := 0; j < n_connectors; j++ {
+		connector_, err := factory.Create(unserializedConfig, log.NewTestLogger(t))
+		assert.NoError(t, err)
+
+		go func(connector deployer.Connector) {
+			for k := 0; k < n_plugin_copies; k++ {
+				for _, testModule_ := range testModules {
+					go func(testModule TestModule) {
+						defer wg.Done()
+
+						output_id, output_data, err := RunStep(
+							t, connector, testModule.location, testModule.stepID, testModule.input)
+						assert.NoError(t, err)
+						assert.Equals(t, output_id, "success")
+						assert.MapNotContainsKeyAny(t, "error", output_data.(map[any]any))
+						if output_id == "error" {
+							errorMsg, ok := output_data.(map[any]any)["error"]
+							if ok {
+								logger.Debugf("plugin error '%s'", errorMsg.(string))
+							}
+						}
+					}(testModule_)
+				}
+			}
+		}(connector_)
+	}
+	// Wait for all the plugins to be done
+	wg.Wait()
+
+	t.Cleanup(func() {
+		assert.NoError(t, os.RemoveAll(rootDir))
+	})
+}
+
+func CreateWorkdir(t *testing.T) string {
+	workdir := fmt.Sprintf("/tmp/%s", RandString(10))
+	if _, err := os.Stat(workdir); !os.IsNotExist(err) {
+		err := os.RemoveAll(workdir)
+		assert.NoError(t, err)
+	}
+	err := os.Mkdir(workdir, os.ModePerm)
+	assert.NoError(t, err)
+	return workdir
+}
```

Comment:
Why are we reinventing this wheel?  Wouldn't something like [`temp.CreateTempDir()`](https://pkg.go.dev/k8s.io/utils/temp#CreateTempDir) be better?

Reply:
maybe, but I'm not touching it during this pr since that would be more related to refactoring the test suite